src/main/java/org/apache/flume/sink/pulsar/PulsarSink.java

```diff
@@ -203,11 +208,28 @@ public Status process() throws EventDeliveryException {
                 } else {
                     counter.incrementBatchCompleteCount();
                 }
-                producer.send(serializeEvent(event, useAvroEventFormat));
+                TypedMessageBuilder<byte[]> newMessage = producer.newMessage();
+                if (event.getHeaders() != null) {
+                    if (event.getHeaders().get("key") != null) {
+                        newMessage = newMessage.key(event.getHeaders().get("key"));
+                    }
+                    newMessage.value(serializeEvent(event, useAvroEventFormat)).properties(event.getHeaders());
+                }
+                if (syncMode) {
+                    newMessage.send();
+                } else {
+                    newMessage.sendAsync().thenAccept(msgId -> {
+                        log.info("Success send message in asyncMode {}", msgId);
```

Comment:
hmm. this log will be printed every message, no?

Reply:
Sorry, i forgot it